sunscreen_runtime/src/debugger/callstack.rs

```diff
@@ -0,0 +1,214 @@
+use backtrace::{Backtrace, BacktraceFrame, SymbolName};
+use radix_trie::Trie;
+use std::collections::HashMap;
+use std::path::Path;
+
+/**
+ * Stores information about individual stack frames.
+ */
+#[derive(Debug, Clone, PartialEq)]
+pub struct StackFrameInfo {
+    /**
+     * Name of the function called.
+     */
+    callee_name: String,
+
+    /**
+     * Name of the file where the callee is defined.
+     */
+    callee_file: String,
+
+    /**
+     * The line number in the file where the callee is defined.
+     */
+    callee_lineno: u32,
+
+    /**
+     * The column index in the file where the callee is defined.
+     */
+    callee_col: u32,
+}
+
+impl StackFrameInfo {
+    /**
+     * Extracts relevant callee information from a `&BacktraceFrame`.
+     */
+    pub fn new(frame: &BacktraceFrame) -> Self {
+        let frame_symbols = frame.symbols();
+        let ip_as_bytes = (frame.ip() as usize).to_ne_bytes();
+        StackFrameInfo {
+            callee_name: frame_symbols[0]
+                .name()
+                .unwrap_or(SymbolName::new(&ip_as_bytes))
+                .to_string(),
+            callee_file: frame_symbols[0]
+                .filename()
+                .unwrap_or(Path::new("No such file"))
+                .to_string_lossy()
+                .into_owned(),
+            callee_lineno: frame_symbols[0].lineno().unwrap_or(0),
+            callee_col: frame_symbols[0].colno().unwrap_or(0),
+        }
+    }
+}
+
+pub trait StackFrames {
```

Comment:
I think this interface should be:
```rust

pub trait IdLookup {
    fn data_to_id(&mut self, key: &[String]) -> u64;
    fn id_to_data(&self, key: u64) -> &[String];
}
```

Then you can abstractly use your trie structure thingy to store both stack traces and group traces.